src/mc_control/mc_global_controller.cpp

```diff
@@ -892,17 +892,23 @@ bool MCGlobalController::run()
     pre_gripper_mbcs_ = controller_->robots().mbcs();
     for(size_t i = 0; i < controller_->robots().size(); ++i)
     {
-      const auto & gi = controller_->robots().robot(i).grippers();
+      auto & robot = controller_->robots().robot(i);
+      auto & realRobot = controller_->realRobot(robot.name());
+      const auto & gi = robot.grippers();
       if(gi.empty())
       {
         continue;
       }
       for(auto & g : gi)
       {
-        g.get().run(controller_->timeStep, controller_->robots().robot(i),
-                    controller_->solver().result().robots_state[i].q);
+        g.get().run(controller_->timeStep, robot, controller_->solver().result().robots_state[i].q);
```

Comment:
I would love to get rid of the `pre_gripper_mbcs_` hack with this PR but I think we need to split `controlToCanonical` in two places for this:
- in the first stage only the control/observed portion is copied to outputRobot/outputRealRobot (copy mbc and sensors only)
- grippers and plugins run on outputRobot
- the customizable part of the converter run to finish the job (enforce mimics by default)

And maybe:
- we keep a `map<string, RobotConverter>` in the global controller (bonus we can have a proper constructor with existing `controlRobot`/`outputRobot`)
- we allow to "mark" some joints to tell the converter they are handled externally (useful for the gripper and for potential plugins)

Ideally I think we should be able to have no gripper joints in the control robot and have the gripper fully handle them

Reply:
Yeah I kept the gripper mechanism as is for now, but I agree it is not nice.
Your proposal sounds good, and fully removing the gripper joints from the QP would definitely be great!

Reply:
I implemented the two steps conversion discussed above and finally got rid of the pre_gripper_mbcs hack :)